mass_mailing_list_dynamic/models/mail_mass_mailing_list.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2017 Tecnativa - Jairo Llopis
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.tools import safe_eval
+
+
+class MassMailingList(models.Model):
+    _inherit = "mail.mass_mailing.list"
+
+    dynamic = fields.Boolean(
+        help="Set this list as dynamic, to make it autosynchronized with "
+             "partners from a given criteria.",
+    )
+    sync_method = fields.Selection(
+        [
+            ("add", "Only add new records"),
+            ("full", "Add and remove records as needed"),
+        ],
+        default="add",
+        required=True,
+        help="Choose the syncronization method for this list if you want to "
+             "make it dynamic",
+    )
+    sync_domain = fields.Char(
+        string="Synchronization critera",
+        default="[('opt_out', '=', False), ('email', '!=', False)]",
+        required=True,
+        help="Filter partners to sync in this list",
+    )
+    is_synced = fields.Boolean(
+        help="Helper field to make the user aware of unsynced changes",
+        default=True,
+    )
+
+    def action_sync(self):
+        """Sync contacts in dynamic lists."""
+        Contact = self.env["mail.mass_mailing.contact"].with_context(
+            syncing=True,
+        )
+        Partner = self.env["res.partner"]
+        # Skip non-dynamic lists
+        dynamic = self.filtered("dynamic")
+        for one in dynamic:
+            sync_domain = safe_eval(one.sync_domain) + [("email", "!=", False)]
```

Comment:
If all times appends "("email", "!=", False)"... you can remove it from default (line 27)

Reply:
Not sure but have a default value will not assure to have this domain as is this is preventing by force in the action

Reply:
Safe domain enforcement is done this way:

```suggestion
            sync_domain = [("email", "!=", False)] + safe_eval(one.sync_domain)
```

It's not like here's a security problem, but it's a good practice. See the security implications here:

![domain](https://user-images.githubusercontent.com/973709/61201925-f3d02b00-a6dd-11e9-9bdb-72d557a5ab9c.png)